### PR TITLE
update python binding samples, shellcode.py with 64bit syscall

### PR DESCRIPTION
**Pull request by xwings**

### CODE REVIEW

File: bindings/python/shellcode.py [hunk 1]
```diff
@@ -93,11 +144,15 @@ def test_i386(mode, code):
...
-        # handle interrupt ourself
-        mu.hook_add(UC_HOOK_INTR, hook_intr)
-
-        # handle SYSCALL
-        mu.hook_add(UC_HOOK_INSN, hook_syscall, None, 1, 0, UC_X86_INS_SYSCALL)
+        if mode == UC_MODE_32:
+            # handle interrupt ourself
+            mu.hook_add(UC_HOOK_INTR, hook_intr)
+        
+        if mode == UC_MODE_32:
```
Comment (aquynh) [inline bindings/python/shellcode.py]: this is in the same block `if mode == UC_MODE_32`, so you can remove this line, and the blank line above

File: bindings/python/shellcode.py [hunk 2]
```diff
@@ -1,27 +1,28 @@
...
+#X86_CODE32 = b"\xeb\x19\x31\xc0\x31\xdb\x31\xd2\x31\xc9\xb0\x04\xb3\x01\x59\xb2\x05\xcd\x80\x31\xc0\xb0\x01\x31\xdb\xcd\x80\xe8\xe2\xff\xff\xff\x68\x65\x6c\x6c\x6f"
+X86_CODE32 = b"\x31\xc0\x50\x68\x2f\x2f\x73\x68\x68\x2f\x62\x69\x6e\x89\xe3\x50\x53\x89\xe1\xb0\x0b\xcd\x80"
 X86_CODE32_SELF = b"\xeb\x1c\x5a\x89\xd6\x8b\x02\x66\x3d\xca\x7d\x75\x06\x66\x05\x03\x03\x89\x02\xfe\xc2\x3d\x41\x41\x41\x41\x75\xe9\xff\xe6\xe8\xdf\xff\xff\xff\x31\xd2\x6a\x0b\x58\x99\x52\x68\x2f\x2f\x73\x68\x68\x2f\x62\x69\x6e\x89\xe3\x52\x53\x89\xe1\xca\x7d\x41\x41\x41\x41\x41\x41\x41\x41"
-
 X86_CODE64 = b"\x48\x31\xff\x57\x57\x5e\x5a\x48\xbf\x2f\x2f\x62\x69\x6e\x2f\x73\x68\x48\xc1\xef\x08\x57\x54\x5f\x6a\x3b\x58\x0f\x05"
 
 # memory address where emulation starts
 ADDRESS = 0x1000000
-
+X86_CODE64_SYSCALL = b'\x0f\x05' # SYSCALL
```
Comment (aquynh) [inline bindings/python/shellcode.py]: is this a dead code?

File: bindings/python/shellcode.py [hunk 3]
```diff
@@ -1,27 +1,28 @@
...
 
 # callback for tracing instructions
 def hook_code(uc, address, size, user_data):
     print(">>> Tracing instruction at 0x%x, instruction size = 0x%x" %(address, size))
     # read this instruction code from memory
     tmp = uc.mem_read(address, size)
-    print("*** EIP = %x *** :" %(address), end="")
+    print("*** XIP = %x *** :" %(address), end="")
```
Comment (aquynh) [inline bindings/python/shellcode.py]: can we change "XIP" to just "PC" ?

File: bindings/python/shellcode.py [hunk 4]
```diff
@@ -65,15 +85,46 @@ def hook_intr(uc, intno, user_data):
...
+
+
+def hook_syscall32(mu, user_data):
+    eax = mu.reg_read(UC_X86_REG_EAX)
+    ebx = mu.reg_read(UC_X86_REG_EBX)
+    print(">>> got SYSCALL with EAX = 0x%x" %(eax))
     mu.emu_stop()
 
+def hook_syscall64(mu, user_data):
+
```
Comment (aquynh) [inline bindings/python/shellcode.py]: please remove this blank line

File: bindings/python/shellcode.py [hunk 5]
```diff
@@ -65,15 +85,46 @@ def hook_intr(uc, intno, user_data):
...
+    rsi = mu.reg_read(UC_X86_REG_RSI)
+    rip = mu.reg_read(UC_X86_REG_RIP)
+
+    print(">>> got SYSCALL with RAX = %d" %(rax))
+    
+    if rax == 59:    #sys_execve
+        #print(">>> rbx=0x%x, rcx=0x%x, rdx=0x%x" % (rbx, rcx, rdx))
+        filename = read_string(mu, rdi)
+        print(">>> SYS_EXECV filename=%s" % filename)
+
```
Comment (aquynh) [inline bindings/python/shellcode.py]: remove this line, too

File: bindings/python/shellcode.py (hunk 2 above)
Comment (xwings) [reply bindings/python/shellcode.py]: nope

File: bindings/python/shellcode.py (hunk 3 above)
Comment (xwings) [reply bindings/python/shellcode.py]: right. fair

File: bindings/python/shellcode.py (hunk 5 above)
Comment (xwings) [reply bindings/python/shellcode.py]: done

File: bindings/python/shellcode.py [hunk 6]
```diff
@@ -30,7 +30,18 @@ def hook_code(uc, address, size, user_data):
 # callback for tracing basic blocks
 def hook_block(uc, address, size, user_data):
     print(">>> Tracing basic block at 0x%x, block size = 0x%x" %(address, size))
+    return
```
Comment (aquynh) [inline bindings/python/shellcode.py]: this return can be removed

File: bindings/python/shellcode.py [hunk 7]
```diff
@@ -65,15 +95,36 @@ def hook_intr(uc, intno, user_data):
...
+
+    print(">>> got SYSCALL with RAX = %d" %(rax))
+    
+    if rax == 59:    #sys_execve
+        filename = read_string(mu, rdi)
+        print(">>> SYS_EXECV filename=%s" % filename)
+
+    else:
+        print("=" * 20)
+        print(">>> Syscall Found 0x%x: interrupt 0x%x, RAX = 0x%x" %(eip, intno, eax))
```
Comment (aquynh) [inline bindings/python/shellcode.py]: this line should be fixed: after copy/paste, we no longer have eip, intno & eax here?

File: bindings/python/shellcode.py [hunk 8]
```diff
@@ -1,27 +1,39 @@
...
+# mov    %esp,%ebx
+# push   %eax
+# push   %ebx
+# mov    %esp,%ecx
+# mov    $0xb,%al
+# int    $0x80
+X86_CODE32 = b"\x31\xc0\x50\x68\x2f\x2f\x73\x68\x68\x2f\x62\x69\x6e\x89\xe3\x50\x53\x89\xe1\xb0\x0b\xcd\x80"
 X86_CODE32_SELF = b"\xeb\x1c\x5a\x89\xd6\x8b\x02\x66\x3d\xca\x7d\x75\x06\x66\x05\x03\x03\x89\x02\xfe\xc2\x3d\x41\x41\x41\x41\x75\xe9\xff\xe6\xe8\xdf\xff\xff\xff\x31\xd2\x6a\x0b\x58\x99\x52\x68\x2f\x2f\x73\x68\x68\x2f\x62\x69\x6e\x89\xe3\x52\x53\x89\xe1\xca\x7d\x41\x41\x41\x41\x41\x41\x41\x41"
-
 X86_CODE64 = b"\x48\x31\xff\x57\x57\x5e\x5a\x48\xbf\x2f\x2f\x62\x69\x6e\x2f\x73\x68\x48\xc1\xef\x08\x57\x54\x5f\x6a\x3b\x58\x0f\x05"
```
Comment (aquynh) [inline bindings/python/shellcode.py]: why you are at it, can you please paste assembly of this x64 shellcode here, as comments?

File: bindings/python/shellcode.py [hunk 9]
```diff
@@ -65,15 +92,36 @@ def hook_intr(uc, intno, user_data):
         print(">>> 0x%x: interrupt 0x%x, EAX = 0x%x" %(eip, intno, eax))
 
 
-def hook_syscall(mu, user_data):
-    rax = mu.reg_read(UC_X86_REG_RAX)
-    print(">>> got SYSCALL with RAX = 0x%x" %(rax))
+def hook_syscall32(mu, user_data):
+    eax = mu.reg_read(UC_X86_REG_EAX)
+    ebx = mu.reg_read(UC_X86_REG_EBX)
```
Comment (aquynh) [inline bindings/python/shellcode.py]: is this dead code?

Comment (xwings) [reply bindings/python/shellcode.py]: done

File: bindings/python/shellcode.py (hunk 7 above)
Comment (xwings) [reply bindings/python/shellcode.py]: updated

File: bindings/python/shellcode.py [hunk 10]
```diff
@@ -61,19 +99,44 @@ def hook_intr(uc, intno, user_data):
...
+    rip = mu.reg_read(UC_X86_REG_RIP)
+
+    print(">>> got SYSCALL with RAX = %d" %(rax))
+    
+    if rax == 59:    #sys_execve
+        filename = read_string(mu, rdi)
+        print(">>> SYS_EXECV filename=%s" % filename)
+
+    else:
+        print("=" * 20)
```
Comment (aquynh) [inline bindings/python/shellcode.py]: to be consistent, please remove this print, and also the print 2 line below

**Comment by aquynh** [comment]
merged, thanks!

**Pull request merged**